File: articles/cell-signaling-communication-biological-coordination/c/signaling_kernel.c
/*
 * Compact signaling kernel in C.
 */

#include <math.h>
#include <stdio.h>

double receptor_occupancy(double L, double Kd) {
    return L / (Kd + L);
}

double hill_response(double L, double K, double n) {
    return pow(L, n) / (pow(K, n) + pow(L, n));
}

double messenger_decay(double M0, double k, double t) {
    return M0 * exp(-k * t);
}

double signaling_half_life(double k) {
    return log(2.0) / k;
}

double quorum_update(double Q, double N, double a, double d, double dt) {
    double next = Q + (a * N - d * Q) * dt;
    if (next < 0.0) {
        return 0.0;
    }
    return next;
}

int main(void) {
    double L = 3.0;
    double Kd = 1.5;
    double K = 2.0;
    double n = 3.0;
    double k = log(4.0) / 4.0;

    printf("occupancy=%.6f\n", receptor_occupancy(L, Kd));
    printf("hill_response=%.6f\n", hill_response(L, K, n));
    printf("signal_at_4_min=%.6f\n", messenger_decay(100.0, k, 4.0));
    printf("half_life_min=%.6f\n", signaling_half_life(k));
    printf("quorum_next=%.6f\n", quorum_update(0.5, 1e8, 1e-9, 0.35, 0.1));

    return 0;
}

File: articles/cell-signaling-communication-biological-coordination/fortran/signaling_kernel.f90
! Compact signaling numerical kernel in Fortran.

program signaling_kernel
  implicit none

  real :: L
  real :: Kd
  real :: K
  real :: n
  real :: occupancy
  real :: hill
  real :: k_decay
  real :: half_life
  real :: signal_4
  real :: Q
  real :: N
  real :: a
  real :: d
  real :: dt
  real :: Q_next

  L = 3.0
  Kd = 1.5
  K = 2.0
  n = 3.0

  occupancy = L / (Kd + L)
  hill = (L ** n) / (K ** n + L ** n)

  k_decay = log(4.0) / 4.0
  half_life = log(2.0) / k_decay
  signal_4 = 100.0 * exp(-k_decay * 4.0)

  Q = 0.5
  N = 1.0e8
  a = 1.0e-9
  d = 0.35
  dt = 0.1
  Q_next = Q + (a * N - d * Q) * dt

  if (Q_next < 0.0) Q_next = 0.0

  print *, "Occupancy:", occupancy
  print *, "Hill response:", hill
  print *, "Signal at 4 min:", signal_4
  print *, "Half-life:", half_life
  print *, "Quorum next:", Q_next
end program signaling_kernel
